fix(server-functions): omitted any from fieldErrors

tzdesign opened the pull request:
Typescript crashes because any could be infinite

fix #6719

# What is it?

- [ ] Feature / enhancement
- [x] Bug
- [ ] Docs / tests / types / typos

# Description

See

- #6719 

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/QwikDev/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] Added new tests to cover the fix / functionality

tzdesign commented:
@gioboa @shairez Looks like I've tripped myself up. We sometimes use any because we have multi-step forms. Typescript unfortunately can't handle it. Showing: `Type instantiation is excessively deep and possibly infinite.`

I rewrote it so that the type never ends up in fieldErrors because Zod ignores any anyway.

I hope one of you can write typescript a little better than I can. I have patched it with the small change to
```{} extends T[K]``` which excludes all any types. Let's see if it still works like this in later Typescript versions.

wmertens commented on starters/apps/qwikcity-test/src/routes/(common)/actions/issue5463/index.tsx:
```diff
@@ -19,6 +19,7 @@ export const useDotNotationAction = globalAction$(
           firstName: z.string(),
         }),
       }),
+      someAny: z.any(), // test for making sure it does not add it to the fieldErrors see #6719
```
I think we need an explicit test for it in the spec file

tzdesign replied:
@wmertens added a test in `server-functions.unit.ts`.

But `expectTypeOf` is pretty weird. I found some issues in vitest and expect-type implementation:
See https://github.com/mmkal/expect-type/issues/31

The any implementation with not seems to work. But I don't know if this will fail at least on build.full.

Let me know what you think.

wmertens commented on packages/qwik-city/runtime/src/types.ts:
```diff
@@ -364,13 +364,15 @@ export type FailOfRest<REST extends readonly DataValidator[]> = REST extends rea
-        : T[K] extends object
-          ? ValidatorErrorKeyDotNotation<T[K], `${Prefix}${K}.`>
-          : `${Prefix}${K}`;
+      [K in keyof T & string]: {} extends T[K]
```
Can you test this type? You can find examples of type tests in the `packages/qwik/src/core/render/jsx/types/jsx-types.unit.tsx` for example.

tzdesign replied:
I will check it out after my short vacation on Thursday.

I think this change will remove the dot notation, but let me test everything. 

The idea of testing the type is perfect, will do it too.

tzdesign commented:
@fabian-hiller just to let you know the ValidatorErrorKeyDotNotation wasn't capturing Record types. Added IsAny to the types.ts in the runtime and added tests.

Maybe you can catch up with me some time talking about rewriting this a bit more.
There are still edge-cases for arrays in my concept, but it's better than having the top-level keys with flatten rod errors.

fabian-hiller commented:
Have you seen my PR #6752?

wmertens commented:
@fabian-hiller I'd like to merge this already, it will cause your pr to get merge conflicts though. Ok?

fabian-hiller commented on packages/qwik-city/src/runtime/src/types.ts:
```diff
@@ -362,17 +362,27 @@ export type FailOfRest<REST extends readonly DataValidator[]> = REST extends rea
-          : `${Prefix}${K}`;
-    }[keyof T & string]
-  : never;
+export type IsAny<T> = 0 extends T & 1 ? true : false;
```
I think this is just a util type similar to the `Prettify` type above and should not be exported.

fabian-hiller commented:
@wmertens and @tzdesign this PR contains the same `any` fix that #6752. To avoid merge conflicts it could be simpler to just move over the `server-functions.unit.ts` file to my PR and add Tobi as a co-author.

tzdesign commented:
@fabian-hiller I'm fine with it, but I copied your type and your type includes any as type string.  So I added a bit more logic.

I'm fine merging the prs together.

fabian-hiller commented:
> I copied your type and your type includes any as type string. So I added a bit more logic.

Can you explain this? Maybe I missed something.

fabian-hiller commented:
Ahh... So you recommend that I also copy your entire `ValidatorErrorKeyDotNotation` type?

wmertens commented:
> Ahh... So you recommend that I should also copy your entire `ValidatorErrorKeyDotNotation` type?

I think that's a good idea

fabian-hiller commented:
Done

tzdesign commented:
@fabian-hiller perfect!

Thank you guys.
